ui/client/components/graph/svg-export/prepareSvg.ts

```diff
@@ -0,0 +1,139 @@
+/* eslint-disable i18next/no-literal-string */
+import * as joint from "jointjs"
+import css from "!raw-loader!./export.styl"
+import {toXml} from "../../../common/SVGUtils"
+import {memoize} from "lodash"
+
+const xmlns = "http://www.w3.org/2000/svg"
+let iconId = 1
+
+function createStyle() {
+  const style = document.createElementNS(xmlns, "style")
+  style.appendChild(document.createTextNode(css))
+  return style
+}
+
+function getCanvas(width: number, height: number) {
+  const canvas = document.createElement("canvas")
+  canvas.width = width
+  canvas.height = height
+  const ctx = canvas.getContext("2d")
+  return {canvas, ctx}
+}
+
+const getPng = memoize((url: string, width: number, height: number) => {
+  const {canvas, ctx} = getCanvas(width, height)
+  return new Promise<[string, string]>(resolve => {
+    const id = `id${iconId++}`
+    const img = new Image(width, height)
+    img.onload = () => {
+      ctx.drawImage(img, 0, 0)
+      resolve([canvas.toDataURL("image/png"), id])
+    }
+    img.src = url
+  })
+})
+
+function getPngScale(url: string, width: number, height: number, scale = 1) {
+  return getPng(url, width * scale, height * scale)
+}
+
+export function debugInWindow(svgString: string) {
+  window.open(null).document.write(svgString)
+}
+
+async function embedImage(image: SVGImageElement): Promise<[SVGImageElement, string]> {
+  const {width, height, href} = image
+  const [dataurl, id] = await getPngScale(href.baseVal, width.baseVal.value, height.baseVal.value, 10)
+  image.setAttribute("xlink:href", dataurl)
+  return [image, id]
+}
+
+function replaceWithDef(svg: SVGSVGElement) {
+  const defs = document.createElementNS(xmlns, "defs")
+  svg.prepend(defs)
+
+  return ([image, id]: [SVGImageElement, string]) => {
+    if (!svg.getElementById(id)) {
+      const def = image.cloneNode() as SVGImageElement
+      def.setAttribute("id", id)
+      defs.append(def)
+    }
+    const use = document.createElementNS(xmlns, "use")
+    use.setAttribute("xlink:href", `#${id}`)
+    image.replaceWith(use)
+  }
+}
+
+async function embedImages(svg: SVGSVGElement) {
+  const nodes = svg.querySelectorAll<SVGImageElement>("image[*|href]")
+  const externalImages = Array.from(nodes)
+    .filter(({href}) => !href.baseVal.startsWith("data:image"))
+    .map(embedImage)
```

Comment:
If you embed images, why not to delete all duplicated svgs from `assets/img/nodes/*`? We have them in `ui/server/src/main/resources/web/static/assets/nodes` too